README.md
[![CRAN checks](https://cranchecks.info/badges/summary/AggregateR)](https://cran.r-project.org/web/checks/check_results_AggregateR.html)
[![](https://www.r-pkg.org/badges/version/AggregateR?color=orange)](https://cran.r-project.org/package=AggregateR)
[![](https://img.shields.io/badge/devel%20version-0.1.1-orange.svg)](https://github.com/AggregateR)
[![](http://cranlogs.r-pkg.org/badges/grand-total/AggregateR?color=blue)](https://cran.r-project.org/package=AggregateR)


# AggregateR

The `Aggregate` function (not to be confounded with `aggregate`) prepares a `data.frame`, `tibble` or `data.table` for merging by computing the sum, mean and variance of all continuous (integer and numeric) variables by a given variable. 
For all categorical variabes (character and factor), it creates dummies and subsequently computes the sum and the mode by a given variable. 
For all Date variables, it computes the recency and duration by a given variable with repsect the an end date variable. 
For computational speed, all the calculations are done with `data.table`. This functions aims at maximum information extraction with a minimum amount of code.

The package also contains faster implementations of the `dummy` and `categories` function (comparable to the same functions in the `dummy` package). When using the `AggregateR` package, the `dummy`-package is deprecated and the internal `dummy` and `categories` functions are superior in terms of speed.

# Installation

To install the package from CRAN: 

```
install.packages('AggregateR')
```

To instal the package from github: 

```
devtools::install_github ('MatthBogaert/AggregateR')
```

# Usage

This code blocks shows how the `Aggregate` function works when confronted with a table with numeric, categorical and Date variables. `Aggregate` accepts a `data.frame`, `tibble` or `data.table` and outputs by default a `data.table`. Note that the 'by' parameter should be a variable in the table 'x' and only accepts a charachter as an input.

```
#Create some data
data <- data.frame(V1=sample(as.factor(c('yes','no')), 200000, TRUE),
                   V2=sample(as.character(c(1,2,3,4,5)),200000, TRUE),
                   V3=sample(1:20000,200000, TRUE),
                   V4=sample(300:1000, 200000, TRUE),
                   V5 = sample(as.Date(as.Date('2014-12-09'):Sys.Date()-1, origin = "1970-01-01"),200000,TRUE),
                   ID=sample(x = as.character(1:4), size = 200000, replace = TRUE))

Aggregate(x=data,by='ID')

## Calculating categorical variables ... 
## Calculating numerical variables ... 
## Calculating date variables ...
## ID V1_no_sum V1_no_mode V1_yes_sum V1_yes_mode V2_1_sum V2_1_mode V2_2_sum V2_2_mode V2_3_sum
## 1:  1     24911          0      25080           1    10006         0     9990         0    10170
## 2:  2     24938          0      25160           1     9985         0    10073         0    10030
## 3:  3     25070          1      24933           0     9845         0     9987         0    10108
## 4:  4     24926          0      24982           1     9923         0     9891         0     9901
## V2_3_mode V2_4_sum V2_4_mode V2_5_sum V2_5_mode    V3_sum   V3_mean   V3_var   V4_sum  V4_mean
## 1:         0     9887         0     9938         0 498324620  9968.287 33440187 32426370 648.6442
## 2:         0     9962         0    10048         0 499201602  9964.502 33370364 32606808 650.8605
## 3:         0     9988         0    10075         0 501006529 10019.529 33208428 32535970 650.6804
## 4:         0     9939         0    10254         0 499350872 10005.427 33285590 32461104 650.4189
## V4_var V5_duration V5_recency
## 1: 40972.02        2172          1
## 2: 41186.23        2172          1
## 3: 40789.41        2172          1
## 4: 41224.02        2172          1
```

As mentioned, the user can also output a tibble for nicer printing. 

```
Aggregate(x=data,by='ID', tibble = TRUE)

## Calculating categorical variables ... 
## Calculating numerical variables ... 
## Calculating date variables ... 
##A tibble: 4 x 23
## ID    V1_no_sum V1_no_mode V1_yes_sum V1_yes_mode V2_1_sum V2_1_mode V2_2_sum V2_2_mode V2_3_sum
## <chr>     <dbl>      <dbl>      <dbl>       <dbl>    <dbl>     <dbl>    <dbl>     <dbl>    <dbl>
##1 1         25060          1      24906           0    10046         0     9847         0     9932
##2 2         25056          1      24964           0     9981         0    10010         0     9986
##3 3         24986          0      25068           1     9989         0    10057         0    10076
##4 4         25037          1      24923           0    10086         0     9955         0    10075
## ... with 13 more variables: V2_3_mode <dbl>, V2_4_sum <dbl>, V2_4_mode <dbl>, V2_5_sum <dbl>,
##   V2_5_mode <dbl>, V3_sum <dbl>, V3_mean <dbl>, V3_var <dbl>, V4_sum <dbl>, V4_mean <dbl>,
##   V4_var <dbl>, V5_duration <dbl>, V5_recency <dbl>
```
# Contact

Compose a friendly e-mail to <Matthias.Bogaert@UGent.Be>.



inst/NEWS
Change log:

Version 0.1.1, Nov 18, 2020
----------------------------
Small bug fix such that functions works when table only cotains numerics or dates. 

Version 0.1.0, Nov 17, 2020
----------------------------
Re-written the code in data.table for computational speed. 
Added efficient dummy-coding also written in data.table.
Allowed the fucntionality to output a tibble. 

Version 0.0.2, Nov 20, 2015
----------------------------
Added the functionality to compute recency and duration of Date variables by an ID

Version 0.0.1, Sep 30, 2015
----------------------------
Submitted package to cran with one function: Aggregate


R/dummy.R
#' Fast-automatic Dummy Variable Creation with Support for Predictive Contexts
#'
#' \code{dummy} creates dummy variables of all the factors and character vectors in a data frame or data table. It also supports settings in which the user only wants to compute dummies for the categorical values that were present in another data set. This is especially useful in the context of predictive modeling, in which the new (test) data has more or other categories than the training data.For computational speed, the code is written in \code{data.table}.
#'
#' @param x a data frame or data table containing at least one factor or character vector
#' @param p Only relevant if object is NULL. Select the top p values in terms of frequency. Either "all" (all categories in all variables), an integer scalar (top p categories in all variables), or a vector of integers (number of top categories per variable in order of appearance).
#' @param object output of the \code{categories} function. This parameter is to be used when dummies should be created only of categories present in another data set (e.g., training set)
#' @param num should the dummies be of class numeric (TRUE) or factor (FALSE). Setting this to TRUE will speed up execution considerably.
#' @param verbose logical. Used to show progress. Does not work when \code{parallel="variable"}.
#' @param ref logical. Only relevant when x is a data.table. If TRUE x will be overwritten by the dummy output (called transformed x), and a reference (i.e., not a copy) to the transformed x will be returned invisibly. If FALSE, x will be left untouched, and the output will be returned as usual. The difference between ref=TRUE and ref=FALSE is that the former uses less memory equal to the amount of the original x (not transformed x). If x=TRUE only the transformed x survives the function. If x=FALSE both the original x and the output (equal in size as transformed x) will survive. The difference is hence the size of the original x, and therefore ref=TRUE is more memory efficient.
#' @examples
#' #create toy data
#' (traindata <- data.frame(var1=as.factor(c("a","b","b","c")),
#'                          var2=as.factor(c(1,1,2,3)),
#'                          var3=c("val1","val2","val3","val3"),
#'                          stringsAsFactors=FALSE))
#' (newdata <- data.frame(var1=as.factor(c("a","b","b","c","d","d")),
#'                        var2=as.factor(c(1,1,2,3,4,5)),
#'                        var3=c("val1","val2","val3","val3","val4","val4"),
#'                        stringsAsFactors=FALSE))
#' #create dummies of training set
#' (dummies_train <- dummy(x=traindata))
#' #create dummies of new set
#' (dummies_new <- dummy(x=newdata))
#'
#' #how many new dummy variables should not have been created?
#' sum(! colnames(dummies_new) %in% colnames(dummies_train))
#'
#' #create dummies of new set using categories found in training set
#' (dummies_new <- dummy(x=newdata,object=categories(traindata,p="all")))
#'
#' #how many new dummy variables should not have be created?
#' sum(! colnames(dummies_new) %in% colnames(dummies_train))
#'
#'
#' #create dummies of training set,
#' #using the top 2 categories of all variables found in the training data
#' dummy(x=traindata,p=2)
#'
#' #create dummies of training set,
#' #using respectively the top 2,3 and 1 categories of the three
#' #variables found in training data
#' dummy(x=traindata,p=c(2,3,1))
#'
#' #create all dummies of training data
#' dummy(x=traindata)
#'
#' \dontrun{
#' #######################
#' #example ref parameter
#'
#' #ref=TRUE, example 1
#' (DT = data.table(a=c("a","b"),b=c("c","c")))
#' dummy(DT,ref=TRUE)
#' DT[] #DT has changed
#'
#' #ref=TRUE, example 2
#' #uses exactly same amount of memory as example 1
#' (DT = data.table(a=c("a","b"),b=c("c","c")))
#' d1 <- dummy(DT,ref=TRUE)
#' DT[] #DT has changed
#' d1[] #d1 is a reference (not a copy) to DT
#'
#' #ref=FALSE, example 3
#' #example 1 and 2 are more memory efficient than example 3
#' (DT = data.table(a=c("a","b"),b=c("c","c")))
#' d2 <- dummy(DT, ref=FALSE)
#' DT[] #DT has not changed
#' d[]
#' # deleting DT after dummy finishes would result in the same final
#' # memory footprint as example 1 and 2, except that in example 3
#' # memory usage is higher when dummy is being executed, and this may be
#' # problematic when DT is large.
#' }
#' @seealso \code{\link{categories}}
#' @return  A data frame or data table containing dummy variables. If ref=TRUE then the output will be invisible and x will contain the output. NOTE: data.table currently has a print bug. In some cases the output does not print. Running the output object multiple times or running it once with [] appended will make it print. In either case, the output will be produced. str() also always works.
#' @author Authors: Michel Ballings, and Dirk Van den Poel, Maintainer: \email{Michel.Ballings@@GMail.com}
#' @export

dummy <- function(x,
                  p="all",
                  object=NULL,
                  num=TRUE,
                  verbose=FALSE,
                  ref=FALSE){


      if(all(class(x)=="data.frame") || ref==FALSE) {
        # for data.frame
        # will break reference if used on data.table
        # (i.e., operate on object by value)
        colnames(x) <- make.names(colnames(x),TRUE)
      } else {
        # for data.table when we want to keep
        # the reference (i.e., operate on object by reference)
        # (note that the solution for data.frame
        # would work too, but it would break the reference)
        setnames(x,names(x),make.names(names(x),TRUE))
      }


      if(is.null(object)) object <- categories(x,p=p)
      ans <- list()

      len <- length(object)

      #error handling
      if (!any(class(x) %in% c("data.table","data.frame"))) stop("x needs to be either a data.frame or data.table")

      #store class of x at input
      if(all(class(x)=="data.frame")) cl <- "data.frame" else cl <- "data.table"

      #change to data.table if x was a data.frame at input
      if (!any(class(x)=="data.table")) setDT(x)

      if (verbose) cat("Start\n")
      for (i in 1:len){

      #for each value an ifelse

      ii <- 0
      envir <- environment()
      for (z in object[[i]]){

          if (num==FALSE) {
            x[, make.names(paste0(names(object)[i],"_",z),TRUE) :=  as.factor(ifelse(get(names(object)[i]) == get("z",envir= envir),1,0))]
          } else if (num==TRUE) {
            x[, make.names(paste0(names(object)[i],"_",z),TRUE) :=  ifelse(get(names(object)[i]) == get("z",envir= envir),1,0)]
          }


        if (verbose) {
            if (ii != 0 && w==z) ii <- 0
            ii <- ii + 1
            cat("   ",round((ii*100)/length(object[[i]]),0),"% of categories processed \n")
          }
        w <- z

      }

      if (verbose) cat(round((i*100)/len,0),"% of variables processed \n")
    }

    #remove original categoricals
    for (i in 1:len){
      x[ , names(object)[i]:= NULL ]
    }

    #change back to data.frame if x was a data.frame at input
    if (cl=="data.frame") setDF(x)
    #use x[] instead of x to make sure it prints
    if (any(class(x)=="data.table") && ref==TRUE){
      return(invisible(x))
    } else {
      return(x)
    }
}


R/categories.R
#' Extraction of Categorical Values as a Preprocessing Step for Making Dummy Variables
#'
#'  \code{categories} stores all the categorical values that are present in the factors and character vectors of a data frame. Numeric and integer vectors are ignored. It is a preprocessing step for the \code{dummy} function. This function is appropriate for settings in which the user only wants to compute dummies for the categorical values that were present in another data set. This is especially useful in predictive modeling, when the new (test) data has more or other categories than the training data.
#'
#' @param x data frame or data table containing factors or character vectors that need to be transformed to dummies. Numerics, dates and integers will be ignored.
#' @param p select the top p values in terms of frequency. Either "all" (all categories in all variables), an integer scalar (top p categories in all variables), or a vector of integers (number of top categories per variable in order of appearance.
#' @examples
#' #create toy data
#' (traindata <- data.frame(var1=as.factor(c("a","b","b","c")),
#'                          var2=as.factor(c(1,1,2,3)),
#'                          var3=c("val1","val2","val3","val3"),
#'                          stringsAsFactors=FALSE))
#' (newdata <- data.frame(var1=as.factor(c("a","b","b","c","d","d")),
#'                        var2=as.factor(c(1,1,2,3,4,5)),
#'                        var3=c("val1","val2","val3","val3","val4","val4"),
#'                        stringsAsFactors=FALSE))
#'
#' categories(x=traindata,p="all")
#' categories(x=traindata,p=2)
#' categories(x=traindata,p=c(2,1,3))
#' @seealso \code{\link{dummy}}
#' @return  A list containing the variable names and the categories
#' @author Authors: Michel Ballings, and Dirk Van den Poel, Maintainer: \email{Michel.Ballings@@GMail.com}
#' @export


categories <- function(x,p="all"){
  colnames(x) <- make.names(colnames(x),TRUE)
  categoricals <- which(sapply(x,function(x) is.factor(x) || is.character(x)))


      if (!any(class(x)=="data.table")){ #if data.frame
          x <- x[,categoricals, drop=FALSE]
      } else if (any(class(x)=="data.table")){ #if data.table
          x <- x[,categoricals, with=FALSE]
      }

      cats <- sapply(1:ncol(x),function(z) {

        if (!any(class(x)=="data.table")){  #if data.frame
            cats <- table(x[,z])
        } else if (any(class(x)=="data.table")){ #if data.table
            cats <- table(x[,z, with=FALSE])
        }
        if(is.numeric(p) && length(p) == 1) {
          names(sort(cats,decreasing=TRUE)[1:if(length(cats) <= p) length(cats) else p])
        } else if (is.numeric(p) && length(p) > 1) {
          names(sort(cats,decreasing=TRUE)[1:if(length(cats) <= p[z]) length(cats) else p[z]])
        } else if (p=="all") {
          names(cats)
        }
      },simplify=FALSE)
      names(cats) <- names(x)
      return(cats)
}


R/Aggregate.R
#' Aggregate numeric, Date and categorical variables
#'
#' The \code{Aggregate} function (not to be confounded with aggregate) prepares a data frame or data table for merging by computing the sum, mean and variance of all continuous (integer and numeric) variables by a given variable. For all categorical variabes (character and factor), it creates dummies and subsequently computes the sum and the mode by a given variable. For all Date variables, it computes the recency and duration by a given variable with repsect the an end date variable. For computational speed, all the calculations are done with \code{data.table}. This functions aims at maximum information extraction with a minimum amount of code.
#'
#' @param x A data frame or data table. Categorical variables have to be of type character or factor and continuous variables have to be of type integer or numeric. Date variables should be in the Date format.
#' @param by A character string specifying the variable on which to aggregate the results. Note that 'by' should be a variable of the table 'x'.
#' @param end_ind A Date object, or something which can be coerced by \code{as.Date(origin, ...)} to such an object. If not specified,  we take the \code{Sys.Date()} as end date.
#' @param format A character string. If not specified, the ISO 8601 international standard which expresses a day "\%Y-\%m-\%d" is taken.
#' @param tibble Should the output be a tibble, data frame or data table? By default, the function returns a data frame or data table depending on the input. To return a tibble, the user must set the tibble = TRUE.
#' @param verbose indicator Used to show the progress.
#' @param object Parameter related to the \code{dummy} function. See ?\code{dummy} for more information.
#' @param p Parameter related to the \code{dummy} function. See ?\code{dummy} for more information.
#'
#' @return A data frame, data table or tibble with the aforementioned variables aggregated by the given ID variables. If the input is a data frame, a data frame is returned else a data table is returned.
#'
#' @author Authors: Matthias Bogaert, Michel Ballings, Dirk Van den Poel, Maintainer: \email{matthias.bogaert@@UGent.be}
#' @examples
#' # Example
#' # Create some data
#' data <- data.frame(V1=sample(as.factor(c('yes','no')), 200000, TRUE),
#'                   V2=sample(as.character(c(1,2,3,4,5)),200000, TRUE),
#'                   V3=sample(1:20000,200000, TRUE),
#'                   V4=sample(300:1000, 200000, TRUE),
#'                   V5 = sample(as.Date(as.Date('2014-12-09'):Sys.Date()-1,
#'                   origin = "1970-01-01"),200000,TRUE),
#'                   ID=sample(x = as.character(1:4), size = 200000, replace = TRUE))
#'
#' Aggregate(x=data,by='ID')
#'
#' # Examples of how to use the object and p argument. See dummy and categories function for details.
#' # Aggregate(x=data,by='ID',object=categories(data))
#' # Aggregate(x=data,by='ID',p=2)
#' @import stats methods data.table NCmisc
#' @importFrom utils tail
#' @importFrom tibble as_tibble
#' @export

# From now on the x parameter needs to be a data frame/data table WITH the by column.
# From now on the by parameter needs to be the name of the by column in quotes
# From now on all the analyses are done in data.table and then transformed to data.frame or data.table
# Allow print out in a tibble by setting tibble = TRUE. Default = FALSE
Aggregate <- function (x, by, end_ind = Sys.Date(), format = '%Y-%m-%d', tibble = FALSE, verbose = TRUE,
                       object = NULL, p = "all") {

  options(stringsAsFactors = FALSE)

  Mode <- function(x) {
    ux <- unique(x)
    ux[which.max(tabulate(match(x, ux)))]
  }

  # problem handling
  if (!any(sapply(by,is.character))) stop('by-variable must be a character!')
  if (class(end_ind) != 'Date') stop('The end_ind should be of class Date')
  if (!any(class(x) %in% c("data.table", "data.frame")))stop("x needs to be either a data.frame or data.table")
  if (all(class(x) == "data.frame")) cl <- "data.frame" else cl <- "data.table"
  if (!any(class(x) == "data.table")) setDT(x)

  ###NEW: Assume x is data table for speed issues

    categoricals <- (sapply(x, is.factor) | sapply(x, is.character)) & colnames(x) != by ###NEW
    if (any(categoricals == TRUE)) {
      if(verbose == TRUE) cat('Calculating categorical variables ... \n')

      if(!is.null(object)) {
        ind <- which(names(object) == by)
        object <- object[-ind]
      }

      dummies_df <- dummy(x = x[,categoricals,with=FALSE],num = TRUE, ref = TRUE,
                          object = object, p = p) ### Works with the new dummy function
      dummies_df <- cbind(x[,by, with = FALSE], dummies_df)
      dummies_df <- dummies_df[,as.list(unlist(lapply(.SD, function(x) list(sum=sum(x),
                                                                              mode=Mode(x))))),
                                 by=eval(by)]###NEW
      names(dummies_df) <- gsub('[.]','_',names(dummies_df))
    }



    numerics <- sapply(x, is.numeric) | sapply (x, is.integer)
    if (any(numerics==TRUE)) {
      if(verbose == TRUE) cat('Calculating numerical variables ... \n')
      numerics <- numerics | colnames(x) == by
      numerics_df <- x[, numerics, with = FALSE]
      numerics_df <- numerics_df[,as.list(unlist(lapply(.SD, function(x) list(sum=sum(x),
                                                                                mean=mean(x),
                                                                                var=var(x))))),
                                   by=eval(by)]
      names(numerics_df) <- gsub('[.]','_',names(numerics_df))

    }

    dates <- sapply(x,function(z) is(z,"Date"))
    if(any(dates == TRUE)) {
      if(verbose == TRUE) cat('Calculating date variables ... \n')
      end_ind <- as.Date(end_ind, format = format)
      dates <- dates | colnames(x) == by
      dates_df <- x[,dates, with = FALSE]
      dates_df <- dates_df[, as.list(unlist(lapply(.SD, function (x) list(duration = end_ind - min(x),
                                                                            recency = end_ind - max(x) )))),
                             by = eval(by)]
      names(dates_df) <- gsub('[.]','_',names(dates_df))
    }


    if (any(dates == TRUE) && any(categoricals == TRUE) && any(numerics == TRUE)) {
      mergelist <- list(dummies_df, numerics_df, dates_df)
      final <- Reduce(function(x,y) merge(x,y, by = by), mergelist)
      if (cl == 'data.frame') final <- data.frame(final)
      else final <- data.table (final)
    } else if (any(categoricals == TRUE) && any(numerics == TRUE)) {
      final <- merge(dummies_df, numerics_df, by = by)
      if (cl == 'data.frame') final <- data.frame(final)
      else final <- data.table (final)
    } else if (any(categoricals == TRUE) && any(dates == TRUE)) {
      final <- merge(dummies_df, dates_df, by = by)
      if (cl == 'data.frame') final <- data.frame(final)
      else final <- data.table (final)
    } else if (any(dates == TRUE) && any(numerics == TRUE)) {
      final <- merge(numerics_df, dates_df, by = by)
      if (cl == 'data.frame') final <- data.frame(final)
      else final <- data.table (final)
    } else if (any(categoricals == TRUE)) {
      if (cl == 'data.frame') final <- data.frame(dummies_df)
      else final <- data.table (dummies_df)
    } else if (any(numerics == TRUE)) {
      if (cl == 'data.frame') final <- data.frame(numerics_df)
      else final <- data.table (numerics_df)
    } else if (any(dates == TRUE)) {
      if (cl == 'data.frame') final <- data.frame(dates_df)
      else final <- data.table (dates_df)
    }

    if(tibble) as_tibble(final) else final
}

